### PR TITLE
Avoid recomputing log in reused CPU pointwise

### LINKED ISSUE #95037: [Inductor] [CPU] Huggingface model AllenaiLongformerBase performance regression > 10% on ww07.4
### 🐛 Describe the bug

There is a performance regression on Huggingface model AllenaiLongformerBase  on ww07.4 [TorchInductor CPU Performance Dashboard](https://github.com/pytorch/pytorch/issues/93531#issuecomment-1434019577)  compare with ww07.2 as bellow:

|  ww07.4   |   |     |   |  ww07.2   |   |     |   | Comp  |     |   |
|  ----  | ----  |  ----  | ----  |  ----  | ----  |  ----  | ----  | ----  |  ----  | ----  |
| batch_size  | speedup | inductor  | eager | batch_size  | speedup | inductor  | eager | speedup ratio | eager ratio  | inductor ratio |
|1	|0.9901	|2.3819678	|2.358386319	|1	|1.1392	|2.0668015	|2.354500269	|0.87	|1	|0.87

Please find sw info from [TorchInductor CPU Performance Dashboard](https://github.com/pytorch/pytorch/issues/93531) 

### Versions

Repro:

```
python -m torch.backends.xeon.run_cpu --core_list 0 --ncores_per_instance 1 benchmarks/dynamo/huggingface.py --performance --float32 -dcpu -n50 --inductor  --no-skip --dashboard --only AllenaiLongformerBase  --cold_start_latency --batch_size 1 --threads 1
```

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #184473

Treat log as a heavy CPU pointwise op for reuse materialization so softmax inputs that include relative-position bias are realized once instead of recomputed across softmax loops. Add a T5-style generated-code regression test.

Fixes #95037
Generated by my agent

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

File: torch/_inductor/ir.py [hunk 1]
```diff
@@ -9691,7 +9691,7 @@ def should_realize_on_reuse(self, users: int) -> bool:
             if is_cpu(self.data):
                 # Heuristic for realizing reused result of heavy ops on cpu
                 opcount = self.data.inner_fn_opcount()
-                heavy_ops = ["exp", "sigmoid"]  # a list of heavy ops
+                heavy_ops = ["exp", "log", "sigmoid"]  # a list of heavy ops
```
Comment (aorenste) [inline torch/_inductor/ir.py]: Should the other `log` variants be included too? `log2`, `log10`, and `log1p` are tracked as distinct ops in `OpsHandler` (see `torch/_inductor/ops_handler.py`), so the membership check `"log" in opcount.used_ops` is exact — it will not match those variants. They have similar transcendental cost to `log`, and a model using e.g. `log1p` for numerical stability would not benefit from this heuristic. If the scope was intentionally narrowed to the T5 motivating case in #95037, fine — but worth a comment here explaining why, since the existing `# a list of heavy ops` comment doesn't justify the choice.

File: test/inductor/test_torchinductor.py [hunk 2]
```diff
@@ -4162,6 +4162,57 @@ def fn(a, b):
...
+
+        actual, code = run_and_get_cpp_code(
+            torch.compile(fn, fullgraph=True), scores, table
+        )
+        expected = fn(expected_scores, table)
+
+        self.assertEqual(actual, expected)
+        self.assertEqual(scores, expected_scores)
+        self.assertEqual(code.count("std::log("), 1)
+
```
Comment (aorenste) [inline test/inductor/test_torchinductor.py:4302]: Low confidence — could you double-check this on a vec-ISA machine? The CPU codegen has two `log` emitters: scalar `std::log(x)` (`cpp.py:865`) and vectorized `x.log()` (`cpp.py:1418`). For the realized kernel here (shape `(16, 16)` = 256 elements), both AVX2 (vec width 8) and AVX-512 (vec width 16) divide evenly, so the scalar tail loop should be empty. Tracing through `codegen_loops` / `codegen_conditions`, an empty tail range returns False before `gen_body()` runs, so the tail kernel body is never emitted. If that's accurate, the realized kernel would contain only `.log()` and `code.count("std::log(")` would be `0`, not `1`, on vec-enabled CPU builds.

Dr-CI is green so this probably works in your environment — but it suggests the test only passes when scalar codegen is in effect (e.g. `ATEN_CPU_CAPABILITY=default` or a non-vec build), and would silently start failing on vec-enabled CI configs. Consider one of:
- assert on `code.count(".log()") + code.count("std::log(") == 1` to be vec-agnostic, or
- force `@torch._inductor.config.patch({"cpp.simdlen": 0})` on the test to pin scalar codegen, or
- assert on a kernel-count proxy instead of the textual op.

**Review by aorenste — APPROVED** [review]

**Comment by jansel** [comment]
My agent says the mergebot failure was caused by the dynamic-shape clone of the new generated-code regression test: dynamic CPU codegen can emit both vector and scalar-tail log call sites for one realized log kernel. I updated the assertion to keep the static exact count while accepting that dynamic vector-plus-tail shape, and reran the static, dynamic-shapes, and codegen-dynamic focused tests plus lintrunner.

File: torch/_inductor/ir.py (hunk 1 above)
Comment (jansel) [reply torch/_inductor/ir.py]: My agent says I addressed this by expanding the CPU heavy-op reuse heuristic from just `log` to `log`, `log1p`, `log2`, and `log10` in `torch/_inductor/ir.py`, so the related logarithm variants get the same materialization treatment.

File: test/inductor/test_torchinductor.py (hunk 2 above)
Comment (jansel) [reply test/inductor/test_torchinductor.py:4302]: My agent says I addressed this by making the generated-code assertion count both scalar `std::log(` and vector `.log()` spellings. The static test still requires exactly one log call site, while the dynamic-shape clone allows the expected vector-plus-tail shape without allowing true duplicated softmax recomputation.

**Comment by jansel** [comment]
My agent says I rebased the PR onto current main to refresh CI after Dr. CI showed checkout/promisor-fetch failures rather than a source test failure. I reran the focused static and dynamic softmax regression tests plus `lintrunner -a` locally, and CI has restarted on the rebased head.

**Comment by jansel** [comment]
My agent says I rebased the PR onto current main to refresh CI after mergebot blocked on an unclassified distributed RNG-state failure in test_random_ops that is unrelated to this Inductor CPU softmax/log change. The rebased head keeps the same two-file diff; I reran lintrunner and the focused static/dynamic softmax regression tests locally, and CI has restarted.

**Comment by jansel** [comment]
@claude are those CI failure related?
